pkg/server/server.go

```diff
@@ -0,0 +1,628 @@
+
+	// start sync loop
+	go s.SyncLoop(s.Options.ctx)
+
+	s.birthCry()
+}
+
+// setInitialized sets s.initialized to 1 if value is true. the set operation is atomic
+func (s *Server) setInitialized(value bool) {
+	var initialized int32
+	if value {
+		initialized = 1
+	}
+	atomic.StoreInt32(&s.initialized, initialized)
+}
+
+// isInitialized checks if all relevant k8s resources have been synced to their latest state
+func (s *Server) isInitialized() bool {
+	return atomic.LoadInt32(&s.initialized) > 0
+}
+
+// birthCry send start event to node object where this server is running
+func (s *Server) birthCry() {
+	klog.Infof("Started multi-networkpolicy-tc")
+	s.Recorder.Eventf(s.NodeRef, v1.EventTypeNormal, "Started", "Started multi-networkpolicy-tc.")
+}
+
+// SyncLoop Waits on Server.Initialized then starts Server.syncRunner.Loop() until context is Done
+func (s *Server) SyncLoop(ctx context.Context) {
+	klog.Infof("SyncLoop waiting for server initialization")
+	_ = wait.PollUntilWithContext(ctx, time.Millisecond*500, func(_ context.Context) (done bool, err error) {
+		return s.isInitialized(), nil
+	})
+	klog.Infof("starting sync runner")
+	s.Sync()
+	s.syncRunner.Loop(ctx.Done())
+}
+
+// NewServer creates a new *Server instance
+func NewServer(o *Options) (*Server, error) {
+	var kubeConfig *rest.Config
+	var err error
+	if len(o.Kubeconfig) == 0 {
+		klog.Info("Neither kubeconfig file nor master URL was specified. Falling back to in-cluster config.")
+		kubeConfig, err = rest.InClusterConfig()
+	} else {
+		kubeConfig, err = clientcmd.NewNonInteractiveDeferredLoadingClientConfig(
+			&clientcmd.ClientConfigLoadingRules{ExplicitPath: o.Kubeconfig},
+			&clientcmd.ConfigOverrides{ClusterInfo: clientcmdapi.Cluster{Server: o.master}},
+		).ClientConfig()
+	}
+	if err != nil {
+		return nil, err
+	}
+
+	if o.podRulesPath != "" {
+		// create pod rules directory if it does not exist
+		if _, err := os.Stat(o.podRulesPath); os.IsNotExist(err) {
+			err = os.Mkdir(o.podRulesPath, 0700)
+			if err != nil {
+				return nil, err
+			}
+		}
+	}
+
+	client, err := clientset.NewForConfig(kubeConfig)
+	if err != nil {
+		return nil, err
+	}
+
+	networkPolicyClient, err := multiclient.NewForConfig(kubeConfig)
+	if err != nil {
+		return nil, err
+	}
+
+	netdefClient, err := netdefclient.NewForConfig(kubeConfig)
+	if err != nil {
+		return nil, err
+	}
+
+	hostname, err := multiutils.GetHostname(o.hostnameOverride)
+	if err != nil {
+		return nil, err
+	}
+
+	eventBroadcaster := record.NewBroadcaster()
+	recorder := eventBroadcaster.NewRecorder(
+		scheme.Scheme,
+		v1.EventSource{Component: "multi-networkpolicy-node", Host: hostname})
+
+	nodeRef := &v1.ObjectReference{
+		Kind:      "Node",
+		Name:      hostname,
+		UID:       types.UID(hostname),
+		Namespace: "",
+	}
+
+	syncPeriod := 30 * time.Second
+	minSyncPeriod := 0 * time.Second
+	burstSyncs := 2
+
+	policyChanges := controllers.NewPolicyChangeTracker()
+	if policyChanges == nil {
```

Comment:
With the current implementation it is impossible that New*ChangeTracker() methods will return nil. Why we need this checks?

Reply:
yep ! removed check